File: scripts/validate-links.mjs
import fs from 'fs-extra';
import glob from 'fast-glob';

const URL_REGEX =
  /github\.com\/PrairieLearn\/PrairieLearn\/(?:tree|blob)\/master\/([-a-zA-Z0-9@:%_+.~#?&//=]*)/;

function filePathFromUrl(url) {
  let path = url.match(URL_REGEX)[1];

  // Some URLs include hashes at the end - strip them off
  path = path.replace(/#.*$/, '');

  // Remove any trailing slashes as well
  path = path.replace(/\/$/, '');

  return path;
}

const files = await glob('**/*.{md,html,json}', { ignore: ['node_modules/**'] });

let error = false;

for (const file of files) {
  const contents = await fs.readFile(file, 'utf8');

  // Extract anything that looks like a PrairieLearn source URL
  const matches = contents.match(new RegExp(URL_REGEX, 'g'));

  if (matches) {
    for (const match of matches) {
      // Extract the file path portion of the URL
      const filePath = filePathFromUrl(match);

      // Check that the file exists in the repo
      if (!(await fs.pathExists(filePath))) {
        error = true;
        console.error(`${file}: ${filePath} does not exist in the repo`);
      }
    }
  }
}

if (error) {
  process.exit(1);
}

File: scripts/validate_links.py
import glob
import os
import re
import sys

URL_REGEX = re.compile(
    r"github\.com/PrairieLearn/PrairieLearn/(?:tree|blob)/master/([-a-zA-Z0-9@:%_+.~#?&//=]*)"
)

files = (
    glob.glob("**/*.md", recursive=True)
    + glob.glob("**/*.html", recursive=True)
    + glob.glob("**/*.json", recursive=True)
)

files = [f for f in files if not f.startswith("node_modules/")]

error = False

for file in files:
    with open(file, encoding="utf8") as f:
        contents = f.read()
    matches = URL_REGEX.findall(contents)
    if matches:
        for path in matches:
            path_clean = re.sub(r"#.*$", "", path)  # Strip hashes
            path_clean = re.sub(r"/$", "", path_clean)  # Remove trailing slash
            if not os.path.exists(path_clean):
                error = True
                print(
                    f"{file}: {path_clean} does not exist in the repo", file=sys.stderr
                )

if error:
    sys.exit(1)
